src/infill/ZigzagConnectorProcessor.cpp

```diff
@@ -0,0 +1,156 @@
+    //    and "use end piece" is enabled
+    // Don't add a zag if:
+    //  - the current zag is NOT an end piece, "skip some zags" is enabled, and the current zag lays in a
+    //    segment which needs to be skipped.
+    // Moreover:
+    //  - if a "connected end pieces" is not enabled and this is an end piece, the last line
+    //    of this end piece will not be added.
+    //
+    // The rules above also apply to how the last part is processed (in polygon finishes)
+    //
+    const bool is_this_endpiece = start_scanline_idx == end_scanline_idx;
+    const bool is_this_connection_even = start_scanline_idx % 2 == 0;
+    bool should_skip_this_connection = false;
+    if (this->skip_some_zags && this->zag_skip_count > 0)
+    {
+        //
+        // Here is an illustration of how the zags will be removed.
+        // ***We use skip_count = 3 as an example:***
+        //
+        // segment numbers:    0     1     2     3     4     5     6     7     8     9     10
+        //    top zags ->   |     |-----|     |--x--|     |-----|     |-----|     |--x--|     |
+        //                  |     |     |     |     |     |     |     |     |     |     |     |
+        // bottom zags ->   |--x--|     |-----|     |-----|     |--x--|     |-----|     |-----|
+        //
+        // LEGEND:  |  :    zigs
+        //          -  :    zags
+        //
+        // Following the line from left to right, we want to remove a zag in every 3. So here we
+        // are removing zags 0, 3, 6, 9, ..., which is (n * 3), where n = 0, 1, 2, ...
+        //
+        // We treat top and bottom zags differently:
+        //   - a bottom zag goes from left -> right. Example: for zag 6, it goes from scanline 6 to 7.
+        //   - a top zag goes from right -> left. Example: for zag 3, it goes from scanline 4 to 3.
+        // We call the start and end scanline indices <start> and <end> separately.
+        // So, to get the results described above, we skip zags in the following way:
+        //   - if direction > 0 (bottom zags: left -> right), skip zags whose <start> mod 3 == 0
+        //   - if direction < 0 (top zags: right -> left), skip zags whose <end> mod 3 == 0
+        //
+        if (direction > 0)
+        {
+            should_skip_this_connection = start_scanline_idx % this->zag_skip_count == 0;
+        }
+        else
+        {
+            should_skip_this_connection = (start_scanline_idx - 1) % this->zag_skip_count == 0;
+        }
+    }
+
+    const bool should_add =
+        (is_this_connection_even && !is_this_endpiece && !should_skip_this_connection) // normal connections that should be added
+        || (this->use_endpieces && is_this_endpiece);  // end piece if it is enabled;
+
+    return should_add;
+}
+
+
+void ZigzagConnectorProcessor::registerScanlineSegmentIntersection(const Point& intersection, int scanline_index, int direction)
```

Comment:
Instead of passing `direction` all the way from infil.cpp you can simply do
```
int direction = end_scanline_idx - start_scanline_idx;
assert((direction == 1 || direction == -1 || direction == 0) && "if the current connection is not an endpiece it should have a direction");
```